src/transformers/models/llama/modeling_llama.py

```diff
@@ -784,8 +894,14 @@ def __init__(self, config: LlamaConfig):
         self.padding_idx = config.pad_token_id
         self.vocab_size = config.vocab_size
 
+        # create attention mask cache that trickles down to each attention layer
+        # so that the attention_mask cache can be shared among layers
+        attention_mask_cache = AttentionMaskCache(is_causal=True)
```

Comment:
Re-iterating on the fact that since it's not passed as an argument to the `Model` changing the default class used in the subsequent layers means we have to re-write this class! Let's also support passing it and default to this class wdyt?  
But if we add it in the model then same goes for the XXForCausalLM, which is why choosing here the class form a config could help

Reply:
I don't follow here really. Do you mean `is_causal=True` should be changed to something like `is_causal=self.causal.is_causal` ?

Reply:
No. What I mean is that you are passing the `attention_mask_cache` to the `LlamaDecoderLayer` from the `LlamaModel`, so if anyone wants to use another class (which is part of the reason why we are going with a class design no?) then they are stuck with this one for the  `LlamaModel` . Unless we plan on this only being an internal class and forcing the attention, I think it makes sense to either:
- pass the class or the instance `attention_mask_cache` to the `LlamaModel` (and `LlamaForCausalLM`)
- init the self.attention_mask_cache using a Mapping. 

We know that sliced attention requires a different logic, same for attention sink, so we know for sure that we'll have more attention classes coming no?

Reply:
After online talk, agree with you, not probable that the attention mask will change and let's restrict the usage. Could support 4d input that will just be return later, and ok for me to keep it in the modeling file for now